[Hub Generated] Publish private branch 'eatkimchee-fist-Microsoft.IoTFirmwareDefense-2023-02-08-preview'

eatkimchee opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/published%2FAzure%2Fazure-rest-api-specs-pr%2Featkimchee-fist-Microsoft.IoTFirmwareDefense-2023-02-08-preview...main/).
## ARM API Information (Control Plane)


Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [x] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month. 
  May 31, 2023
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  May 31, 2023
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

eatkimchee committed:
add or modify files

eatkimchee commented:
The PR is created based on the updates in the private branch. The updates in the PR has already been reviewed and approved with this PR [Azure/azure-rest-api-specs-pr/12395](https://github.com/Azure/azure-rest-api-specs-pr/pull/12395)

ihsanali01 reviewed (approved):
Approved

rflmk committed:
Fix validation errors

- remove references to old private preview versions from readme.md
- add binary hardening flags to custom-words.txt

eatkimchee commented:
Thanks @ihsanali01 , just added minor changes that should fix the validation errors (no swagger changes, just readme and custom words). No idea why the SwaggerAPIView check failed the second time, seems unrelated

eatkimchee commented:
I triggered the APIView job to run again and it passed, but this PR didn't update showing that it passed
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2767030&view=results

JackTn commented:
approved @eatkimchee  Do you need merge this pr now ?

eatkimchee commented:
Yes @JackTn  please merge it so we can begin the SDK release process. Thanks!

TimLovellSmith commented:
/azp run

TimLovellSmith commented on custom-words.txt:
```diff
@@ -2810,4 +2810,8 @@ Hnsw
+relro
+RELRO
+runpath
+rpath
```
Hm. Without having yet read the jsons, I have to wonder, are these words? Do we need them to be?

TimLovellSmith replied:
I think it would make more sense to add them under cSpell than in custom-words, if they're just domain-specific terms for your RP, let me show you how...

In [https://github.com/Azure/azure-rest-api-specs/blob/main/cSpell.json](https://github.com/Azure/azure-rest-api-specs/blob/main/cSpell.json)

look at this as an example [where they had to keep a typo, or non-word, for backCompat]:

        {
            "filename": "**/specification/keyvault/data-plane/Microsoft.KeyVault/**/keyvault.json",
            "words": [
                "Regenerte"
            ]
        },

eatkimchee replied:
these 3 are domain specific to our firmware analysis, yes. so i would add an entry to cSpell.json and remove them from custom words?

TimLovellSmith replied:
Yup, that's the idea

TimLovellSmith commented on specification/fist/resource-manager/Microsoft.IoTFirmwareDefense/preview/2023-02-08-preview/firmwares.json:
```diff
@@ -0,0 +1,1983 @@
+          "description": "The rpath of the uploaded firmware.",
+          "type": "string",
+          "x-nullable": true
+        }
```
What's an 'rpath'? Could it be named something else?

eatkimchee replied:
It's specific to our security analysis, it's a known term in the firmware industry

TimLovellSmith commented on specification/fist/resource-manager/Microsoft.IoTFirmwareDefense/preview/2023-02-08-preview/firmwares.json:
```diff
@@ -0,0 +1,1983 @@
+          "type": "string",
+          "x-nullable": true
+        },
+        "runpath": {
```
>runpath [](http://example.com/codeflow?start=9&length=7)

I'd suggest naming this 'runPath' in camel case, instead. Then it won't look misspelled and we won't need to ad it to custom-words.

eatkimchee replied:
same as above, it's known to be spelled like this for linux binary compilation parameters. If it's an issue we could change it to camel case but would prefer not to

TimLovellSmith replied:
You can go with cSpell in that case.

TimLovellSmith commented on specification/fist/resource-manager/Microsoft.IoTFirmwareDefense/preview/2023-02-08-preview/common.json:
```diff
@@ -0,0 +1,186 @@
+          "type": "string",
+          "x-nullable": true
+        }
+      },
```
Could add format: uri

TimLovellSmith commented on specification/fist/resource-manager/Microsoft.IoTFirmwareDefense/preview/2023-02-08-preview/common.json:
```diff
@@ -0,0 +1,186 @@
+  },
+  "paths": {},
+  "definitions": {
+    "UrlToken": {
```
>UrlToken [](http://example.com/codeflow?start=5&length=8)

If they are SAS uris, you should document such for clarity, and mark them as x-ms-secret.

eatkimchee replied:
We used to mark them as x-ms-secret in our swagger, but in a previous review in azure-rest-api-specs-pr, they told us to remove it because that was only used for marking inputs as secret (body parameters), not responses. how should we proceed?

eatkimchee replied:
as discussed we will leave it out for now until the documentation is clear about whether this property applies to responses as well as requests

TimLovellSmith commented on specification/fist/resource-manager/Microsoft.IoTFirmwareDefense/preview/2023-02-08-preview/common.json:
```diff
@@ -0,0 +1,186 @@
+    }
+  },
+  "parameters": {
+    "FirmwareGroupNameParameter": {
```
>FirmwareGroupNameParameter [](http://example.com/codeflow?start=5&length=26)

I don't see it being used

eatkimchee replied:
thanks, can remove for this api version

TimLovellSmith commented on specification/fist/resource-manager/Microsoft.IoTFirmwareDefense/preview/2023-02-08-preview/workspaces.json:
```diff
@@ -0,0 +1,477 @@
+      "in": "body",
+      "required": true,
+      "schema": {
+        "$ref": "#/definitions/GenerateUploadUrlProperties"
```
>GenerateUploadUrlProperties [](http://example.com/codeflow?start=31&length=27)

naming nit, I might call it 'Request' instead of Properties since its just the whole request body/object, and not just the properties, or a set of nested properties

rflmk committed:
make changes from review comments

- remove unused param
- rename definition to match usage
- make sas token response properties more clear with description and format and readOnly
- move domain-specific terms to cSpell.json

rflmk committed:
fix urls in examples

eatkimchee commented:
@TimLovellSmith @JackTn can you tell why the service api readiness test is failing now when it didn't fail for earlier commits? is this blocking?

eatkimchee commented:
/azp run

JackTn commented:
> @TimLovellSmith @JackTn can you tell why the service api readiness test is failing now when it didn't fail for earlier commits? is this blocking?

@konrad-jamrozik  help to tell us this api readiness test failing ? In my opinion this is due to the new api version not release yet. Let me know if i'm wrong. Thanks.

eatkimchee commented:
/azp run

TimLovellSmith commented on specification/fist/resource-manager/Microsoft.IoTFirmwareDefense/preview/2023-02-08-preview/common.json:
```diff
@@ -0,0 +1,182 @@
+          "type": "string",
+          "format": "uri",
+          "readOnly": true,
+          "x-nullable": true
```
> "x-nullable": true [](http://example.com/codeflow?start=9&length=19)

As a general rule we don't recommend using x-nullable, as you can just omit the property from your payload, unless you *actually need to return 'null' from your API for some reason*.

TimLovellSmith replied:
In this case, what does it mean when you return null?

eatkimchee replied:
Ah I see. if omission is allowed that satisfies the backwards compatibility so I will remove the nullable property

rflmk committed:
remove x-nullable from string url properties

eatkimchee commented:
Hi @TimLovellSmith , does this have all the changes you requested now?

TimLovellSmith commented:
I still have this nagging feeling that not marking it as x-ms-secret is going to come back to haunt me... aside from that I don't see any reason to hold this up.

konrad-jamrozik commented:
> > @TimLovellSmith @JackTn can you tell why the service api readiness test is failing now when it didn't fail for earlier commits? is this blocking?
> 
> @konrad-jamrozik help to tell us this api readiness test failing ? In my opinion this is due to the new api version not release yet. Let me know if i'm wrong. Thanks.

Sorry for the late reply!

I see the failure is in `~[Staging] ServiceAPIReadinessTest`. Overall, all failing checks that have `~[Staging]` in the name can be safely ignored for the purpose of determining if the PR can be merged. If `API Readiness Check` would fail we should deliberate, but I see it is passing.

eatkimchee commented:
Thanks @TimLovellSmith and @konrad-jamrozik.  @JackTn whenever you get a chance to merge, it looks like we pass all the necessary checks now. Thanks!

JackTn commented:
@msyyc  can you help to merge this pr, Thanks

msyyc commented:
@eatkimchee Do you want to merge it now?

eatkimchee commented:
@msyyc Yes please merge it now, thanks!
